TechnicalDocuments/Proposals/Fish_Operator.md

```diff
@@ -0,0 +1,95 @@
+# Introduce "fish" operator
+
+* Author: Ilya Puchka
+* Review Manager:TBD
+
+## Introduction
+
+This proposal it to suggest to use new operator that will allow us to compose functions returning optional values, which is not currently possible.
+
+## Motivation
+
+Since introduction of `>>>` operator it was widely adopted throughout the code base and became an important tool in our tool set. But there are some cases where it's not enough. 
+
+One of these cases is when first item returns optional, which makes us to jump through the hoops:
+
+```swift
+didTap: helpText.map { ^Action.didTapWhatDoesThisMean($0) >>> observer }
+```
+
+## Proposed solution
+
+To handle such case [Prelude](https://github.com/pointfreeco/swift-prelude) defines "fish" operator (with a help of free curried `flatMap` function):
+
+```swift
+infix operator >=>: ForwardComposition
+
+public func >=> <A, B, C>(lhs: @escaping (A) -> B?, rhs: @escaping (B) -> C?) -> (A) -> C? {
+  return lhs >>> flatMap(rhs)
+}
+
+public func flatMap<A, B>(_ a2b: @escaping (A) -> B?) -> (A?) -> B? {
+  return { a in
+    a.flatMap(a2b)
+  }
+}
```

Comment:
I am having trouble understanding this `flatMap`, since this doesn't resemble any `flatMap` I have seen. Could you point me to a reference where this is used (outside Swift for example)

Reply:
That's one defined in Prelude. We can make it private if it's an issue or use closure in-place. I don't have examples from outside Swift.

Reply:
Here is another example from Prelude of the similar function applied to any monoid (and Void is monoid as well, same as `Prelude.Unit`)

```
public func foldMap<A, M: Monoid>(_ f: @escaping (A) -> M) -> (A?) -> M {
  return { xs in
    xs.foldMap(f)
  }
}
```